File: assets/manuscript-sync.js
/* AUREFOLD — browser-side manuscript sync.
 * The selected Markdown file is parsed and hashed locally. Only structural
 * metadata + SHA-256 hashes are sent to Supabase; prose never leaves browser.
 */
(function () {
  "use strict";

  const cfg = window.AUREFOLD_COMMUNITY || {};
  const TOKEN_KEY = "aurefold_loremaster_access_token";
  const parserVersion = "aurefold-manuscript-sync-v1";
  const $ = (id) => document.getElementById(id);
  const book = $("lm-sync-book");
  const version = $("lm-sync-version");
  const file = $("lm-sync-file");
  const previewBtn = $("lm-sync-preview-btn");
  const registerBtn = $("lm-sync-register-btn");
  const confirmBox = $("lm-sync-confirm");
  const status = $("lm-sync-status");
  const current = $("lm-sync-current");
  const tbody = $("lm-sync-preview");
  let prepared = null;

  if (!book || !version || !file || !previewBtn || !registerBtn || !confirmBox || !status || !current || !tbody) return;

  function token() { return sessionStorage.getItem(TOKEN_KEY) || ""; }
  async function request(path, options = {}) {
    const accessToken = token();
    if (!accessToken) throw new Error("Not signed in");
    const response = await fetch(`${cfg.supabaseUrl}/rest/v1/${path}`, {
      ...options,
      headers: {
        apikey: cfg.supabaseKey,
        Authorization: `Bearer ${accessToken}`,
        Accept: "application/json",
        ...(options.body ? { "Content-Type": "application/json" } : {}),
        ...(options.headers || {})
      }
    });
    if (!response.ok) {
      const detail = await response.text();
      throw new Error(`Canon API ${response.status}: ${detail.slice(0, 240)}`);
    }
    const text = await response.text();
    if (!text) return null;
    try { return JSON.parse(text); } catch (_) { return text; }
  }

  function normalizeNewlines(text) { return text.replace(/\r\n/g, "\n").replace(/\r/g, "\n"); }
  function normalizeBlock(text) {
    const rows = normalizeNewlines(text).split("\n").map((line) => line.replace(/[ \t]+$/g, ""));
    return rows.join("\n").trim() + "\n";
  }
  function normalizeHeading(text) {
    return String(text || "").normalize("NFKC").trim().toLocaleLowerCase("en")
      .replace(/[‘’]/g, "'").replace(/[“”]/g, '"').replace(/[–—]/g, "-").replace(/\s+/g, " ");
  }
  async function digest(text) {
    const buf = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(text));
    return Array.from(new Uint8Array(buf), (b) => b.toString(16).padStart(2, "0")).join("");
  }
  function wordCount(text) { return (text.match(/[\p{L}\p{N}_’'-]+/gu) || []).length; }
  function sectionType(label) {
    const low = label.toLowerCase();
    if (low.startsWith("chapter")) return "chapter";
    if (low === "interlude") return "interlude";
    if (low === "prologue") return "prologue";
    if (low === "epilogue") return "epilogue";
    return "other";
  }
  function esc(value) { return encodeURIComponent(value); }

  async function loadCurrentSections() {
    const rows = await request(`author_manuscript_current_sections?select=stable_key,section_type,ordinal,chapter_number,heading,content_sha256,body_sha256,word_count&book_code=eq.${esc(book.value)}&order=ordinal.asc`);
    return Array.isArray(rows) ? rows : [];
  }

  async function refreshCurrent() {
    try {
      const rows = await request(`author_manuscript_sync_status?select=book_code,version_label,revision_status,section_count,word_count,pending_reviews&book_code=eq.${esc(book.value)}`);
      const row = Array.isArray(rows) ? rows[0] : null;
      current.textContent = row
        ? `${row.version_label || "No manuscript baseline"} · ${row.revision_status || ""} · ${row.section_count || 0} sections · ${row.pending_reviews || 0} pending reviews`
        : "No manuscript baseline registered.";
    } catch (error) {
      current.textContent = `Could not read sync status: ${error.message}`;
    }
  }

  async function parseMarkdown(text, baseline) {
    const raw = normalizeNewlines(text);
    const lines = raw.split("\n");
    const starts = [];
    const sectionRe = /^##\s+(Chapter\b.*|Interlude|Prologue|Epilogue)\s*$/i;
    for (let i = 0; i < lines.length; i += 1) if (sectionRe.test(lines[i])) starts.push(i);
    if (!starts.length) throw new Error("No Markdown sections found. Expected headings such as '## Chapter One'.");
    starts.push(lines.length);

    const byHeading = new Map();
    const byBody = new Map();
    for (const row of baseline) {
      const hk = normalizeHeading(row.heading);
      if (!byHeading.has(hk)) byHeading.set(hk, []);
      byHeading.get(hk).push(row);
      if (!byBody.has(row.body_sha256)) byBody.set(row.body_sha256, []);
      byBody.get(row.body_sha256).push(row);
    }

    const used = new Set();
    const sections = [];
    let chapterCounter = 0;

    for (let index = 0; index < starts.length - 1; index += 1) {
      const start = starts[index];
      const end = starts[index + 1];
      const label = lines[start].replace(/^##\s+/, "").trim();
      const type = sectionType(label);
      let chapterNumber = null;
      if (type === "chapter") { chapterCounter += 1; chapterNumber = chapterCounter; }

      let heading = "";
      let bodyStart = start + 1;
      const titleMatch = (lines[start + 1] || "").match(/^###\s+(.+?)\s*$/);
      if (titleMatch) { heading = titleMatch[1].trim(); bodyStart = start + 2; }

      const bodyText = normalizeBlock(lines.slice(bodyStart, end).join("\n"));
      const sectionText = normalizeBlock(lines.slice(start, end).join("\n"));
      const [bodyHash, contentHash] = await Promise.all([digest(bodyText), digest(sectionText)]);

      let stableKey = null;
      let mappingStatus = "unmapped";
      const marker = (lines[start - 1] || "").match(/^<!--\s*aurefold:section\s+([A-Za-z0-9._:-]+)\s*-->\s*$/);
      if (marker && !used.has(marker[1])) { stableKey = marker[1]; mappingStatus = "marker"; }
      if (!stableKey && heading) {
        const candidates = (byHeading.get(normalizeHeading(heading)) || []).filter((r) => !used.has(r.stable_key));
        if (candidates.length === 1) { stableKey = candidates[0].stable_key; mappingStatus = "heading"; }
      }
      if (!stableKey) {
        const candidates = (byBody.get(bodyHash) || []).filter((r) => !used.has(r.stable_key));
        if (candidates.length === 1) { stableKey = candidates[0].stable_key; mappingStatus = "body_hash"; }
      }
      if (!stableKey) stableKey = `${book.value.replace("book-", "b")}-new-${contentHash.slice(0, 12)}`;
      if (used.has(stableKey)) throw new Error(`Duplicate stable section key: ${stableKey}`);
      used.add(stableKey);

      sections.push({
        stable_key: stableKey,
        section_type: type,
        chapter_number: chapterNumber,
        ordinal: index + 1,
        label,
        heading,
        start_line: start + 1,
        end_line: end,
        word_count: wordCount(bodyText),
        content_sha256: contentHash,
        body_sha256: bodyHash,
        metadata: { mapping_status: mappingStatus }
      });
    }

    return {
      parser_version: parserVersion,
      source_sha256: await digest(raw),
      section_count: sections.length,
      word_count: sections.reduce((n, s) => n + s.word_count, 0),
      sections
    };
  }

  function compare(baseline, parsed) {
    const oldMap = new Map(baseline.map((r) => [r.stable_key, r]));
    const newMap = new Map(parsed.sections.map((r) => [r.stable_key, r]));
    const keys = [...new Set([...oldMap.keys(), ...newMap.keys()])];
    return keys.map((key) => {
      const a = oldMap.get(key); const b = newMap.get(key);
      let change = "unchanged";
      if (!a) change = "added";
      else if (!b) change = "removed";
      else if (a.body_sha256 !== b.body_sha256) change = "modified";
      else if ((a.heading || "") !== (b.heading || "")) change = "renamed";
      else if (Number(a.ordinal) !== Number(b.ordinal)) change = "moved";
      return { stable_key: key, change_type: change, old: a, next: b, needs_review: change !== "unchanged" };
    }).sort((x, y) => Number((x.next || x.old || {}).ordinal || 9999) - Number((y.next || y.old || {}).ordinal || 9999));
  }

  function renderPreview(changes) {
    tbody.textContent = "";
    for (const row of changes) {
      const tr = document.createElement("tr");
      const vals = [row.next?.ordinal ?? row.old?.ordinal ?? "—", row.stable_key,
        row.next?.heading || row.old?.heading || "—", row.change_type,
        row.next?.metadata?.mapping_status || (row.old ? "baseline" : "—")];
      for (const value of vals) { const td = document.createElement("td"); td.textContent = String(value); tr.appendChild(td); }
      if (row.needs_review) tr.dataset.review = "true";
      tbody.appendChild(tr);
    }
  }

  async function prepare() {
    if (!file.files[0]) throw new Error("Choose a Markdown manuscript file first.");
    if (!version.value.trim()) throw new Error("Enter the new manuscript version label.");
    const baseline = await loadCurrentSections();
    if (!baseline.length) throw new Error("No current manuscript baseline exists for this book.");
    const text = await file.files[0].text();
    const parsed = await parseMarkdown(text, baseline);
    const changes = compare(baseline, parsed);
    const unmapped = parsed.sections.filter((s) => s.metadata.mapping_status === "unmapped").length;
    const review = changes.filter((c) => c.needs_review).length;
    prepared = { baseline, parsed, changes, filename: file.files[0].name };
    renderPreview(changes);
    status.textContent = `${parsed.section_count} sections · ${parsed.word_count} sync-parser words · ${review} changed/added/removed · ${unmapped} unmapped. Prose remained local.`;
    registerBtn.disabled = false;
    confirmBox.checked = false;
  }

  async function register() {
    if (!prepared) throw new Error("Preview the manuscript first.");
    if (!confirmBox.checked) throw new Error("Confirm that this version should become the current operational manuscript baseline.");
    const payload = {
      p_book_code: book.value,
      p_version_label: version.value.trim(),
      p_source_filename: prepared.filename,
      p_source_sha256: prepared.parsed.source_sha256,
      p_word_count: prepared.parsed.word_count,
      p_sections: prepared.parsed.sections,
      p_notes: "Registered through Loremaster Manuscript Sync. Hash-only upload; prose remained local."
    };
    registerBtn.disabled = true;
    status.textContent = "Registering hashes and building review queue…";
    const result = await request("rpc/author_register_manuscript_version", { method: "POST", body: JSON.stringify(payload) });
    status.textContent = `Registered ${payload.p_version_label}. Sync run: ${Array.isArray(result) ? JSON.stringify(result) : result || "complete"}. Review queue created for affected sections.`;
    prepared = null;
    await refreshCurrent();
  }

  previewBtn.addEventListener("click", () => {
    status.textContent = "Parsing and hashing locally…";
    registerBtn.disabled = true;
    prepare().catch((error) => { prepared = null; tbody.textContent = ""; status.textContent = error.message; });
  });
  registerBtn.addEventListener("click", () => register().catch((error) => { registerBtn.disabled = false; status.textContent = error.message; }));
  book.addEventListener("change", () => { prepared = null; tbody.textContent = ""; registerBtn.disabled = true; refreshCurrent(); });
  file.addEventListener("change", () => { prepared = null; registerBtn.disabled = true; confirmBox.checked = false; });

  window.addEventListener("aurefold:loremaster-ready", refreshCurrent);
  if (token()) refreshCurrent();
})();

File: tools/manuscript_sync.py
#!/usr/bin/env python3
"""Aurefold manuscript sync manifest generator.

Reads Markdown locally, hashes chapter/interlude bodies, preserves stable section
identities when possible, and emits a hash-only JSON payload suitable for the
Supabase `author_register_manuscript_version` RPC.

The manuscript prose is never included in the output payload.
"""
from __future__ import annotations

import argparse
import hashlib
import json
import re
import sys
import unicodedata
from pathlib import Path
from typing import Any

PARSER_VERSION = "aurefold-manuscript-sync-v1"
SECTION_RE = re.compile(r"^##\s+(Chapter\b.*|Interlude|Prologue|Epilogue)\s*$", re.I)
TITLE_RE = re.compile(r"^###\s+(.+?)\s*$")
MARKER_RE = re.compile(r"^<!--\s*aurefold:section\s+([A-Za-z0-9._:-]+)\s*-->\s*$")
WORD_RE = re.compile(r"\b[\w’'-]+\b", re.UNICODE)


def normalize_newlines(text: str) -> str:
    return text.replace("\r\n", "\n").replace("\r", "\n")


def normalize_block(text: str) -> str:
    text = normalize_newlines(text)
    return "\n".join(line.rstrip() for line in text.split("\n")).strip() + "\n"


def sha256(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def normalize_heading(text: str) -> str:
    text = unicodedata.normalize("NFKC", text).strip().casefold()
    text = text.translate(str.maketrans({"’": "'", "‘": "'", "“": '"', "”": '"', "–": "-", "—": "-"}))
    return re.sub(r"\s+", " ", text)


def section_type(label: str) -> str:
    low = label.casefold()
    if low.startswith("chapter"):
        return "chapter"
    if low == "interlude":
        return "interlude"
    if low == "prologue":
        return "prologue"
    if low == "epilogue":
        return "epilogue"
    return "other"


def load_baseline(path: Path | None) -> dict[str, Any]:
    if not path:
        return {"sections": []}
    return json.loads(path.read_text(encoding="utf-8"))


def parse_markdown(path: Path, book_code: str, baseline: dict[str, Any]) -> dict[str, Any]:
    raw = normalize_newlines(path.read_text(encoding="utf-8"))
    lines = raw.splitlines()
    starts: list[int] = []
    for i, line in enumerate(lines):
        if SECTION_RE.match(line):
            starts.append(i)
    if not starts:
        raise ValueError("No Markdown chapter/interlude headings found (expected '## Chapter ...' / '## Interlude').")
    starts.append(len(lines))

    baseline_sections = baseline.get("sections", [])
    by_heading: dict[str, list[dict[str, Any]]] = {}
    by_body: dict[str, list[dict[str, Any]]] = {}
    for row in baseline_sections:
        by_heading.setdefault(normalize_heading(row.get("heading", "")), []).append(row)
        by_body.setdefault(row.get("body_sha256", ""), []).append(row)

    used_keys: set[str] = set()
    sections: list[dict[str, Any]] = []
    chapter_counter = 0

    for ordinal, (start, end) in enumerate(zip(starts[:-1], starts[1:]), 1):
        label = lines[start][3:].strip()
        stype = section_type(label)
        if stype == "chapter":
            chapter_counter += 1
            chapter_number: int | None = chapter_counter
        else:
            chapter_number = None

        heading = ""
        body_start = start + 1
        if start + 1 < end:
            m_title = TITLE_RE.match(lines[start + 1])
            if m_title:
                heading = m_title.group(1).strip()
                body_start = start + 2

        body_text = normalize_block("\n".join(lines[body_start:end]))
        section_text = normalize_block("\n".join(lines[start:end]))
        body_hash = sha256(body_text)
        content_hash = sha256(section_text)

        marker_key = None
        if start > 0:
            m_marker = MARKER_RE.match(lines[start - 1])
            if m_marker:
                marker_key = m_marker.group(1)

        stable_key = None
        mapping_status = "unmapped"
        if marker_key:
            stable_key = marker_key
            mapping_status = "marker"
        else:
            candidates = by_heading.get(normalize_heading(heading), []) if heading else []
            candidates = [c for c in candidates if c.get("stable_key") not in used_keys]
            if len(candidates) == 1:
                stable_key = candidates[0]["stable_key"]
                mapping_status = "heading"
            else:
                body_candidates = [c for c in by_body.get(body_hash, []) if c.get("stable_key") not in used_keys]
                if len(body_candidates) == 1:
                    stable_key = body_candidates[0]["stable_key"]
                    mapping_status = "body_hash"

        if not stable_key:
            prefix = book_code.replace("book-", "b")
            stable_key = f"{prefix}-new-{content_hash[:12]}"
        if stable_key in used_keys:
            raise ValueError(f"Duplicate stable_key produced: {stable_key}")
        used_keys.add(stable_key)

        sections.append({
            "stable_key": stable_key,
            "section_type": stype,
            "chapter_number": chapter_number,
            "ordinal": ordinal,
            "label": label,
            "heading": heading,
            "start_line": start + 1,
            "end_line": end,
            "word_count": len(WORD_RE.findall(body_text)),
            "content_sha256": content_hash,
            "body_sha256": body_hash,
            "metadata": {"mapping_status": mapping_status},
        })

    return {
        "parser_version": PARSER_VERSION,
        "file": path.name,
        "source_sha256": sha256(raw),
        "word_count": sum(s["word_count"] for s in sections),
        "section_count": len(sections),
        "sections": sections,
    }


def compare(baseline: dict[str, Any], current: dict[str, Any]) -> list[dict[str, Any]]:
    old = {s["stable_key"]: s for s in baseline.get("sections", [])}
    new = {s["stable_key"]: s for s in current.get("sections", [])}
    rows: list[dict[str, Any]] = []
    for key in sorted(set(old) | set(new), key=lambda k: (new.get(k) or old.get(k)).get("ordinal", 999999)):
        a, b = old.get(key), new.get(key)
        if a is None:
            kind = "added"
        elif b is None:
            kind = "removed"
        else:
            modified = a.get("body_sha256") != b.get("body_sha256")
            moved = a.get("ordinal") != b.get("ordinal")
            renamed = a.get("heading") != b.get("heading")
            if modified:
                kind = "modified"
            elif renamed:
                kind = "renamed"
            elif moved:
                kind = "moved"
            else:
                kind = "unchanged"
        rows.append({
            "stable_key": key,
            "change_type": kind,
            "old_heading": a.get("heading") if a else None,
            "new_heading": b.get("heading") if b else None,
            "old_ordinal": a.get("ordinal") if a else None,
            "new_ordinal": b.get("ordinal") if b else None,
            "mapping_status": (b or {}).get("metadata", {}).get("mapping_status"),
            "needs_review": kind != "unchanged",
        })
    return rows


def make_payload(parsed: dict[str, Any], book_code: str, version_label: str, notes: str | None) -> dict[str, Any]:
    return {
        "p_book_code": book_code,
        "p_version_label": version_label,
        "p_source_filename": parsed["file"],
        "p_source_sha256": parsed["source_sha256"],
        "p_word_count": parsed["word_count"],
        "p_sections": parsed["sections"],
        "p_notes": notes,
    }


def write_tagged_copy(source: Path, parsed: dict[str, Any], target: Path) -> None:
    lines = normalize_newlines(source.read_text(encoding="utf-8")).splitlines()
    by_line = {s["start_line"] - 1: s["stable_key"] for s in parsed["sections"]}
    out: list[str] = []
    for i, line in enumerate(lines):
        if i in by_line:
            if not (out and MARKER_RE.match(out[-1])):
                out.append(f"<!-- aurefold:section {by_line[i]} -->")
        out.append(line)
    target.write_text("\n".join(out) + "\n", encoding="utf-8")


def main() -> int:
    ap = argparse.ArgumentParser(description="Generate Aurefold hash-only manuscript sync payloads.")
    ap.add_argument("manuscript", type=Path)
    ap.add_argument("--book-code", default="book-1")
    ap.add_argument("--version-label", required=True)
    ap.add_argument("--baseline", type=Path)
    ap.add_argument("--out", type=Path, required=True, help="Write RPC payload JSON here")
    ap.add_argument("--comparison-out", type=Path, help="Optional change preview JSON")
    ap.add_argument("--tagged-out", type=Path, help="Optional Markdown copy with stable section markers")
    ap.add_argument("--notes")
    args = ap.parse_args()

    baseline = load_baseline(args.baseline)
    parsed = parse_markdown(args.manuscript, args.book_code, baseline)
    changes = compare(baseline, parsed) if baseline.get("sections") else []
    payload = make_payload(parsed, args.book_code, args.version_label, args.notes)
    args.out.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    if args.comparison_out:
        args.comparison_out.write_text(json.dumps(changes, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    if args.tagged_out:
        write_tagged_copy(args.manuscript, parsed, args.tagged_out)

    counts: dict[str, int] = {}
    for row in changes:
        counts[row["change_type"]] = counts.get(row["change_type"], 0) + 1
    unmapped = sum(1 for s in parsed["sections"] if s["metadata"]["mapping_status"] == "unmapped")
    print(f"{parsed['file']}: {parsed['section_count']} sections, {parsed['word_count']} sync-parser words")
    print(f"source sha256: {parsed['source_sha256']}")
    if changes:
        print("changes:", ", ".join(f"{k}={v}" for k, v in sorted(counts.items())))
    print(f"unmapped sections: {unmapped}")
    print(f"payload: {args.out}")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:
        print(f"error: {exc}", file=sys.stderr)
        raise SystemExit(2)
